components/SocketIOProvider.tsx

```diff
@@ -0,0 +1,27 @@
+import React, { ReactNode, useEffect, useState } from "react"
+import { io, Socket } from "socket.io-client"
+
+export const SocketIOContext = React.createContext<Socket>(null as any)
+
+interface SocketIOProviderProps {
+  children: ReactNode
+}
+
+export default function SocketIOProvider({ children }: SocketIOProviderProps) {
+  const [socket, setSocket] = useState<Socket>(null as any)
+
+  useEffect(() => {
+    fetch("/api/socketio-init").finally(() => {
+      setSocket((socket) => {
+        socket = io()
+        return socket
+      })
+    })
+  }, [])
```

Comment:
If you had gone with this approach, we would only run this once (on app load).  You'd want to handle to errors and connection drops, then retry.